src/Main.java

```diff
@@ -1,54 +1,32 @@
 import java.util.Random;
-import java.util.Scanner;
 
 public class Main {
 
-	static final int CODE_LENGTH = 4;
-	static final int MAX_SECRET = 10000;
-
 	public static void main(String[] args) {
-		// Set up code validation
-		Validator codeValidator = new Validator(CODE_LENGTH, false);
-
-		// Create secret
-		Random randomGenerator = new Random();
-		int secretKey = 0;
-		boolean validatedSecret = false;
-		while (!validatedSecret) {
-			secretKey = randomGenerator.nextInt(MAX_SECRET);
-			validatedSecret = codeValidator.isValidCode(String.valueOf(secretKey));
-		}
-		System.out.println("The secret is " + secretKey);
-		Digits secret = (Digits) CodeFactory.getInstance(secretKey);
-
-		Scanner input = new Scanner(System.in);
-
-		boolean gameIsWon = false;
-
-		while (!gameIsWon) {
-			// Get guesses from the user
-			System.out.println("Guess a 4-digit number: ");
-
-			String userInput = input.next();
-
-			if (codeValidator.isValidCode(userInput)) {
-				Digits guess = (Digits) CodeFactory.getInstance(Integer.parseInt(userInput));
-
-				if (guess.equals(secret)) {
-					gameIsWon = true;
-					System.out.println("Game should be finished");
-				} else {
-					System.out.println("Guess does not match the secret. Please try again.");
-				}
-			} else {
-				System.out.println("Invalid guess. Please try again.");
-			}
-			;
-
+		Reader reader = new Reader();
+		Printer printer = new Printer();
+
+		Secret secret = generateGameSecret();
+		printer.printInstruction("The secret is " + secret.getDigits());
+
+		Checker checker = new Checker(secret);
+		
+		printer.printInstruction("Please enter a 4 digit guess: ");
+		Secret guess = reader.readUserSecret();
+
+		while (!secret.equals(guess)) {
```

Comment:
This is a **very small** but easy opportunity to grant other developers clarity of our intent.
By lifting this logic into a meaningful concept within our domain we can help others understand what is more obvious to those of us with context.

ie:

`
private boolean gameOver(Secret guess, Secret secret) {
    return secret.equals(guess);
}`

`while(!gameOver()) {...}`

Methods are free! 
 
Now the concept of "The game ends when the guess is identical to the secret" is spelt out by this method.
If a newcomer cares to better understand what the conditions for a game being over are as they read this, they can chose to investigate themselves - or not!

It's a courtesy I'm always grateful to receive.